_posts/2015-09-01-Cocoa-Architecture:-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,146 @@
+
+        if ([cell isSelected]) {
+            [ARAnalytics event:ARRemoveFromAlbumEvent withProperties:@{
+                @"artworks" : @(self.artworks.count),
+                @"from" : [ARNavigationController pageID]
+            }];
+            [...]
+```
+
+We would instead build something [like this](https://github.com/artsy/eigen/blob/master/Artsy/App/ARAppDelegate+Analytics.m#L69):
+
+``` objc
+@implementation ARAppDelegate (Analytics)
+
+- (void)setupAnalytics
+{
+    ArtsyKeys *keys = [[ArtsyKeys alloc] init];
+	[...]
+    [ARAnalytics setupWithAnalytics: @{ [...] } configuration:
+    @{
+        ARAnalyticsTrackedEvents:
+            @[
+                @{
+                    ARAnalyticsClass: ARAddToAlbumViewController.class,
+                    ARAnalyticsDetails: @[
+                        @{
+                            ARAnalyticsEventName: ARRemoveFromAlbumEvent,
+                            ARAnalyticsSelectorName: NSStringFromSelector(@selector(tableView: didSelectRowAtIndexPath:)),
+                            ARAnalyticsProperties: ^NSDictionary*(ARAddToAlbumViewController *controller, NSArray *_) {
+                                return @{
+                                    @"artworks" : @(controller.artworks.count),
+                                    @"from" : [ARNavigationController pageID],
+                                };
+                        },
+                        [...]
+                    ]
+                },
+				[...]
+```
+
+This gives us the ability to not sprinkle analytics code around the app in every file. It keeps the responsibilities of objects simpler and we've been happy with it in Eigen. We've not migrated it into Energy, it's dependency on ReactiveCocoa brings too much additional weight. If you want to learn more about this pattern check out [Aspect-Oriented Programming and ARAnalytics](http://artsy.github.io/blog/2014/08/04/aspect-oriented-programming-and-aranalytics/).
+
+## Class Methods as the whole API
+
+For a very long time, I preferred aesthetics of class based APIs. I think I still do. However, once you start adding tests to a project then they become a bit of a problem.
+
+I'm a big fan of the idea of Dependency Injection within tests. This, roughly _TL:DR'd_, means passing in any additional context instead of an object creating the context itself. A common case is a call to `NSUserDefaults`. It's very likely _not_ the role of your class to know which `NSUserDefault`s object you're working with, but it's likely that you're making that decision in the method itself by doing something like `[[NSUserDefaults standardUserDefaults] setObject:reminderID forKey:@"ARReminderID"];`. Using dependency injection would be allowing that object to come from outside that method. If you're interested in a longer explanation, read this great [objc.io](https://www.objc.io/issues/15-testing/dependency-injection/) by [Jon Reid](http://qualitycoding.org/about/).
```

Comment:
"passing in any additional context instead of an object creating the context itself" -> "passing any additional context into an object, instead of having the object create the context itself" (Maybe, tripped me up on the first read.)

The example of NSUserDefaults is great, since it's approachable for beginners, but it doesn't completely jive with the tl;dr you provided (since they don't actually create NSUserDefaults, they just access a shared state). Might not be a big deal. Could change "you're working with, but it's" you're working with (`standardUserDefaults` or otherwise), but it's" ?

It's a tricky concept to explain in passing – good idea to link to other resources. Actually a DI article could be a good idea...